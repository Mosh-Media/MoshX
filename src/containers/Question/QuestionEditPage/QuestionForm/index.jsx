import React, { lazy } from "react";

import Button from "@material-ui/core/Button";
import { useForm } from "react-hook-form";
import styles from "./styles";
import { withRouter } from "react-router-dom";
import withStyles from "@material-ui/core/styles/withStyles";
import { compose } from "recompose";
import MoPageHeaderEdit from "components/library/MoPageHeaderEdit";
import MoSpinner from "components/library/MoSpinner";
import { useCallback } from "react";
import MoHintEdit from "components/library/MoHintEdit";
import MoPageContentEdit from "components/library/MoPageContentEdit";
import MoTypography from "components/library/MoTypography";

const CodeEditor = lazy(() => import("components/shared/CodeEditor"));

const QuestionForm = ({
  classes,
  children,
  isLoading,
  label,
  question,
  subtitle,
  setQuestion,
  title,
  navToQuestionViewPage,
  onSubmit
}) => {
  const { handleSubmit, register } = useForm();

  const handleChange = useCallback(({ userAnswer, key }) => {
    if (userAnswer === "{}" || userAnswer === "") {
      return;
    }
    setQuestion({ ...question, [key]: userAnswer });
  },
    [setQuestion, question]
  );

  console.log(question)
  if (isLoading || !question) {
    return <MoSpinner isLoading={isLoading} color="primary" />;
  }
  return (
    <form onSubmit={handleSubmit(e => onSubmit({ ...question, ...e }))}>
      <section className={classes.section}>
        <MoPageHeaderEdit
          text={title}
          register={register}
          name="title"
          placeholder={"Html Paragraph"}
        />
        <MoPageContentEdit
          text={label}
          register={register}
          name="label"
          placeholder={"Create a paragraph open tag in html"}
        />
        <MoHintEdit
          text={subtitle}
          register={register}
          name="subtitle"
          placeholder={"<p>"}
        />
      </section>
      {question && (
        <section className={classes.section}>
          <MoTypography
            color="grey"
            font="breeSerif"
            marginBottom="sm"
            variant="h6"
          >
            Question
          </MoTypography>
          <CodeEditor
            codeAnswer={"Write Question Here"}
            codeLanguage={question?.language}
            codeQuestion={question?.question}
            isConsole={true}
            isEditMode={true}
            noInline={false}
            isPlayground={question?.isPlayground}
            handleOnChange={userAnswer => handleChange(userAnswer, "question")}
            sm={6}
            md={6}
          />
        </section>
      )}
      {question && (
        <section className={classes.section}>
          <MoTypography
            color="grey"
            font="breeSerif"
            marginBottom="sm"
            variant="h6"
          >
            Answer
          </MoTypography>
          <CodeEditor
            codeAnswer={"Answer"}
            codeLanguage={question?.language}
            codeQuestion={question?.answer}
            isConsole={true}
            isEditMode={true}
            noInline={false}
            isPlayground={question?.isPlayground}
            handleOnChange={userAnswer => handleChange(userAnswer, "answer")}
            sm={6}
            md={6}
          />
        </section>
      )}

      {children && <section className={classes.section}>{children}</section>}
      {question && (
        <section className={classes.section}>
          <Button
            type="button"
            color="default"
            onClick={() => navToQuestionViewPage()}
          >
            Back to Question
          </Button>
          <Button
            onKeyPress={e => {
              e.key === "Enter" && e.preventDefault();
            }}
            type="submit"
            color="primary"
          >
            Save
          </Button>
        </section>
      )}
    </form>
  );
};

export default compose(withRouter, withStyles(styles))(QuestionForm);
