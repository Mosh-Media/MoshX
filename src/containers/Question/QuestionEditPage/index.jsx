/**
 *
 * @author MoSkool
 * @version 1.0.0
 * @visibleName Question Edit Mode Component ‚ÅâÔ∏è
 *
 * Single Question is code editor, code preview and error console. This container fetches a single question
 * This is the edit mode of the question, where edits can happen and only admins can access this for now
 *
 * @param {Object} firebase - Firebase class provides access to authUser and db - comes from withAuthentication hoc
 * @param {Object} match - Contains information about how a <Route path> matched the URL - comes from withRouter and passed to withAuthentication hoc
 * @param {Object} history - Provides several different implementations for managing session history in JavaScript in various environments - comes from withRouter and passed to withAuthentication hoc
 * @withAuthentication - HOC provides firebase and match props
 * @returns {<QuestionForm/>} - returns CodeEditor component which renders the rest of the components
 *
 * @see Link [Example Question Page in Edit Mode](https://moskool.com/courses/mo-easy/1/isEditMode)
 */

import React, { useCallback, useEffect, useState, Suspense } from "react";

import * as ROLES from "constants/roles";

import MoSnackbar from "components/library/MoSnackBar";
import MoSpinner from "components/library/MoSpinner";
import QuestionForm from "containers/Question/QuestionEditPage/QuestionForm";
import withAuthorization from "components/shared/Session/withAuthorization";
import { compose } from "recompose";
import { withAuthentication } from "components/shared/Session";
import { updateQuestion } from "utils/questionFirebase";

const QuestionEditPage = ({ authUser, firebase, history, match }) => {
  const [isLoading, setIsLoading] = useState(true);
  const [question, setQuestion] = useState();
  const [snackbarProps, setSnackbarProps] = useState(null);

  const viewQuestion = useCallback(() => {
    history.push(
      `/courses/${match.params.collection}/${match.params.questionId}`
    );
  }, [history, match]);

  const onSubmit = useCallback(
    (event, formData) => {
      updateQuestion({ ...formData, ...event }, firebase, match);
      setSnackbarProps({
        autoHideDuration: 2000,
        buttonText: "View Question",
        isActive: true,
        title: "Saved"
      });
    },
    [firebase, setSnackbarProps, match]
  );

  /* TODO: Move to custom hook */
  useEffect(() => {
    const id = match.params.questionId;
    setIsLoading(true);
    const unsubscribe = firebase
      .getCollectionById(`courses/${match.params.collection}/questions`, id)
      .onSnapshot(snapshot => {
        if (snapshot.size) {
          const question = [];
          snapshot.forEach(doc =>
            question.push({ ...doc.data(), uid: doc.id })
          );
          try {
            /* Questions can contain special JSON characters that needs to be parsed */
            setQuestion({
              ...question[0],
              question: JSON.parse(question[0].question)
            });
          } catch {
            setQuestion(question[0]);
          }
        } else if (id === "new") {
          setQuestion({
            title: "Title goes here",
            label: "Subtitle goes here",
            question: "<h1>Question goes here</h1>",
            answer: "<h1>Answer goes hereüéâ</h1>",
            language: "html"
          });
        }
        setIsLoading(false);
      });

    return () => {
      unsubscribe();
      setSnackbarProps(null);
    };
  }, [firebase, match]);

  if (!match.params && !question) {
    return;
  }

  return (
    <Suspense fallback={<MoSpinner isLoading color="primary" />}>
      <QuestionForm
        isLoading={isLoading}
        isCard={false}
        title={question?.title}
        label={question?.label}
        question={question}
        setQuestion={e => setQuestion(e)}
        subtitle={question?.subtitle}
        viewQuestion={() => viewQuestion()}
        onSubmit={(event, formData) => onSubmit(event, formData)}
      />
      {!isLoading && snackbarProps && (
        <MoSnackbar authUser={authUser} snackbarProps={snackbarProps} />
      )}
    </Suspense>
  );
};

const condition = authUser => authUser && !!authUser.roles[ROLES.ADMIN];

export default compose(
  withAuthentication,
  withAuthorization(condition)
)(QuestionEditPage);
