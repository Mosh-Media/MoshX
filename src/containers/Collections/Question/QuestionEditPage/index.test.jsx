import React from "react";

import ShallowRenderer from "react-test-renderer/shallow";
import QuestionEditPage from ".";

const renderer = new ShallowRenderer();

describe("QuestionEditPage Component", () => {
  test("should match the snapshot", () => {
    renderer.render(<QuestionEditPage />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
