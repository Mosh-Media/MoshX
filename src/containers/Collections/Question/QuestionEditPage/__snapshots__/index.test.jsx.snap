// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionEditPage Component should match the snapshot 1`] = `
<ForwardRef(WithStyles)
  maxWidth="xl"
>
  <ForwardRef(WithStyles)
    Breadcrumbs={[Function]}
  />
  <withRouter(WithStyles(QuestionForm))
    isCard={false}
    isLoading={true}
    navToQuestionViewPage={[Function]}
    onSubmit={[Function]}
    question={null}
    setQuestion={[Function]}
  />
</ForwardRef(WithStyles)>
`;
